### ISSUE TITLE
Proposal: Switch from strftime to internal implementation for locale independent formats

### ISSUE THREAD

**Issue by phprus** [fmtlib/fmt#2541]
Implementation of the idea:
> I'm not sure if we should add more special cases but it would be nice to get rid of `strftime` completely in `formatter::format` by parsing the format string and using the optimization when possible. It would be not as fast as special casing but would be more generally applicable.
>
> _Originally posted by @vitaut in https://github.com/fmtlib/fmt/issues/2500#issuecomment-920078450_

### My implementation
https://github.com/phprus/fmt/tree/optimize-tm-formatting-3

Functions https://github.com/phprus/fmt/blob/6c13ec65d06c5c0bba7806a01252179ae76cb1c9/include/fmt/chrono.h#L1545-L1576
can be optimized.

### Performance
Not tested yet.

### Problem
I don't know why, but on Windows the ``detail::write`` ( https://github.com/phprus/fmt/blob/6c13ec65d06c5c0bba7806a01252179ae76cb1c9/include/fmt/chrono.h#L364-L369 ) function works with a different locale than ``strftime``.

```
 3/19 Test  #3: chrono-test ......................***Failed    0.06 sec
[ RUN      ] chrono_test.time_point
D:\a\fmt\fmt\test\chrono-test.cc(179): error: Expected equality of these values:
  output
    Which is: "Sat Oct  9 19:37:48 2021"
  fmt::format(fmt::runtime(fmt_spec), t1)
    Which is: "10/09/21 19:37:48"
D:\a\fmt\fmt\test\chrono-test.cc(180): error: Expected equality of these values:
  output
    Which is: "Sat Oct  9 19:37:48 2021"
  fmt::format(fmt::runtime(fmt_spec), tm)
    Which is: "10/09/21 19:37:48"
D:\a\fmt\fmt\test\chrono-test.cc(179): error: Expected equality of these values:
  output
    Which is: "Sat Oct  9 19:37:48 2021"
  fmt::format(fmt::runtime(fmt_spec), t1)
    Which is: "10/09/21 19:37:48"
D:\a\fmt\fmt\test\chrono-test.cc(180): error: Expected equality of these values:
  output
    Which is: "Sat Oct  9 19:37:48 2021"
  fmt::format(fmt::runtime(fmt_spec), tm)
    Which is: "10/09/21 19:37:48"
D:\a\fmt\fmt\test\chrono-test.cc(179): error: Expected equality of these values:
  output
    Which is: "07:37:48 PM"
  fmt::format(fmt::runtime(fmt_spec), t1)
    Which is: "19:37:48"
D:\a\fmt\fmt\test\chrono-test.cc(180): error: Expected equality of these values:
  output
    Which is: "07:37:48 PM"
  fmt::format(fmt::runtime(fmt_spec), tm)
    Which is: "19:37:48"
[  FAILED  ] chrono_test.time_point (3 ms)


16/19 Test #16: xchar-test .......................***Failed    0.03 sec
[ RUN      ] chrono_test.time_point
D:\a\fmt\fmt\test\xchar-test.cc(289): error: Expected equality of these values:
  output
    Which is: L"Sat Oct  9 19:37:49 2021"
  fmt::format(fmt_spec, t1)
    Which is: L"10/09/21 19:37:49"
D:\a\fmt\fmt\test\xchar-test.cc(290): error: Expected equality of these values:
  output
    Which is: L"Sat Oct  9 19:37:49 2021"
  fmt::format(fmt_spec, tm)
    Which is: L"10/09/21 19:37:49"
D:\a\fmt\fmt\test\xchar-test.cc(289): error: Expected equality of these values:
  output
    Which is: L"Sat Oct  9 19:37:49 2021"
  fmt::format(fmt_spec, t1)
    Which is: L"10/09/21 19:37:49"
D:\a\fmt\fmt\test\xchar-test.cc(290): error: Expected equality of these values:
  output
    Which is: L"Sat Oct  9 19:37:49 2021"
  fmt::format(fmt_spec, tm)
    Which is: L"10/09/21 19:37:49"
D:\a\fmt\fmt\test\xchar-test.cc(289): error: Expected equality of these values:
  output
    Which is: L"07:37:49 PM"
  fmt::format(fmt_spec, t1)
    Which is: L"19:37:49"
D:\a\fmt\fmt\test\xchar-test.cc(290): error: Expected equality of these values:
  output
    Which is: L"07:37:49 PM"
  fmt::format(fmt_spec, tm)
    Which is: L"19:37:49"
[  FAILED  ] chrono_test.time_point (3 ms)

```

**Comment by vitaut** [comment]
`strftime` uses the C locale while `detail::write` uses the C++ locale (`std::locale`). We should use the latter and update the tests.

**Comment by phprus** [comment]
C and C++ locales are not compatible in Windows? I was unable to get the same output for ``strftime`` and ``detail::write``.
I think I need help with work with ```std::locale```...

I plan to write performance tests in the next few days.

Now all locale independent formats are implemented without ``strftime``.

**Comment by phprus** [comment]
``libstdc++`` in ``gcc-4`` does not support ``std::put_time``.

**Comment by phprus** [comment]
Other prototype for fast time formatting from @toughengineer (https://github.com/toughengineer/fmt, https://github.com/toughengineer/fmt/commit/3ae5d81627a82fe0f05e83bc57a44cf034f4c83b)

**Comment by toughengineer** [comment]
If you plan to stop using `strftime`,
* you'll have to remove mentions of `strftime` (to set users' expectations straight) and fully specify supported formatting features, and
* **fmtlib** will stop being a drop-in replacement for `strftime` in some cases, e.g. see here: https://godbolt.org/z/8Wz79YeaT.

What's the policy regarding compatibility with `strftime`?

**Comment by phprus** [comment]
These cases are GNU extensions. POSIX and MSVC ``strftime`` doesn't support them (https://pubs.opengroup.org/onlinepubs/9699919799/functions/strftime.html, https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strftime-wcsftime-strftime-l-wcsftime-l?view=msvc-140)

The C and C++ standard defines a limited set of supported formats.

**Comment by toughengineer** [comment]
To clarify my earlier comment, currently the documentation [says](https://fmt.dev/latest/syntax.html#chrono-specs):
> `std::tm` uses the system’s `strftime` so refer to its documentation for details on supported conversion specifiers.

which means that I should be able to use whatever format features supported by "the system’s `strftime`".

**Comment by phprus** [comment]
@vitaut, what do you think about this?
Need to maintain compatibility with C/C++ standard or POSIX or GNU or MS?

**Comment by vitaut** [comment]
1. We don't need to support non-standard format specs. Giving an error is perfectly fine.
2. No, we shouldn't fallback to a single call because it may give inconsistent results. Each format spec should be processed individually.
3. Separate please.

> Maybe there is an error in this code and conversion to UTF-8 should be performed only for char?

`wchar_t` is not implemented there yet, feel free to change.

**Comment by toughengineer** [comment]
If non-standard specs are not needed to be supported and each format spec should be processed individually, why not use locale facet instead of `strftime` and kill two birds with one stone also supporting `format()` calls with locale?

**Comment by toughengineer** [comment]
Is speed in `FMT_COMPILE` case as important as in the runtime case? I sped up formatting with arbitrary compile time format string roughly twice, should it be added in this library? If yes, should the speed be achieved by _any_ means, or is there a limit to complexity of the code?

**Comment by toughengineer** [comment]
@vitaut
> Definitely and thank you for doing this. Are you referring to #2543?

No, I made a PoC and was wondering if that's something of interest for this library.

> There is a limit, we probably don't want to go very platform-specific.

Not platform specific, no JIT, just pure C++ (hopefully even without UB).

**Comment by vitaut** [comment]
PR is welcome!
